### DISCUSSION TITLE
Provide an adapter for collections

### DISCUSSION THREAD

**Discussion by pBouillon** [state-adapt/state-adapt#81] [Ideas]
Hi,

StateAdapt features some adapters out of the box for the primitive types, but collections are not part of them.

If my store features an array of something, I need to either define an adapter for this array or rely on the entity adapter, I don't think there is an in-between.

Would it be possible to have a `createArrayAdapter` or equivalent (`createCollectionAdapter`, etc.) for such types?

It could help reducing the duplicated code between projects as well as ease the adoption of StateAdapt with out-of-the-box essentials.

I also believe that collections are not always used as entities, and be able to treat them as such could result in a more straightforward way to declare and manipulate them as part of our state.

Once again, thanks for your work on this library, it's refreshing to see a declarative alternative to state management out there!

Here is the idea, with a little more code to explicit it:

## Current State

```ts
import { createAdapter, joinAdapters } from '@state-adapt/core';
import { EntityState, createEntityState } from '@state-adapt/core/adapters';

export interface TodoItem {
  id: number;
  title: string;
  completed: boolean;
}

export const todoItemAdapter = createAdapter<TodoItem>()({/*...*/});

// 👇 Require this "intermediary" adapter
export const todoItemsAdapter = createEntityAdapter<TodoItem>()(todoItemAdapter);

export interface TodoItemsState {
  pagination: Pagination;
  // 👇 The state can't be expressed without StateAdapt's types
  todoItems: EntityState<TodoItem, 'id'>;
}

export const todoItemsStateAdapter = joinAdapters<TodoItemsState>()({
  pagination: paginationAdapter,
  todoItems: todoItemsAdapter,
})();

const initialState: Readonly<TodoItemsState> = {
  pagination: { /*...*/ },
  // 👇 Creating the initial state with a simple empty array is not possible
  todoItems: createEntityState<TodoItem, 'id'>(),
};
```

## Suggested Feature

```diff
import { createAdapter, joinAdapters } from '@state-adapt/core';
- import { EntityState, createEntityState } from '@state-adapt/core/adapters';
+ import { createArrayAdapter } from '@state-adapt/core/adapters';

export interface TodoItem {
  id: number;
  title: string;
  completed: boolean;
}

export const todoItemAdapter = createAdapter<TodoItem>()({/*...*/});

- export const todoItemsAdapter = createEntityAdapter<TodoItem>()(todoItemAdapter);

export interface TodoItemsState {
  pagination: Pagination;
-  todoItems: EntityState<TodoItem, 'id'>;
+ todoItems: TodoItem[];
}

export const todoItemsStateAdapter = joinAdapters<TodoItemsState>()({
  pagination: paginationAdapter,
-  todoItems: todoItemsAdapter,
+ todoItems: createArrayAdapter<TodoItem>(),
})();

const initialState: Readonly<TodoItemsState> = {
  pagination: { /*...*/ },
- todoItems: createEntityState<TodoItem, 'id'>(),
+ todoItems: [],
};
```

**Comment by pBouillon** [comment]
Following up the suggestion, here is a possible implementation:

```ts
import { createAdapter } from '@state-adapt/core';

export const createCollectionAdapter = <TItem>() =>
  createAdapter<TItem[]>()({
    add: (state: TItem[], item: TItem) => [...state, item],
    concat: (state: TItem[], items: TItem[]) => state.concat(items),
    insert: (state: TItem[], [item, index]: [TItem, number]) => {
      const isIndexWithinBounds = index >= 0 && index < state.length;
      return isIndexWithinBounds
        ? [...state.slice(0, index), item, ...state.slice(index)]
        : state;
    },
    clear: () => [],
    removeAll: (state: TItem[], item: TItem) => state.filter((i: TItem) => i !== item),
    removeAt: (state: TItem[], index: number) => {
      const isIndexWithinBounds = index >= 0 && index < state.length;
      return isIndexWithinBounds
        ? [...state.slice(0, index), ...state.slice(index + 1)]
        : state;
    },
    selectors: {
      isEmpty: (state: TItem[]) => state.length === 0,
      length: (state: TItem[]) => state.length,
    },
  });
```

Not sure about the interface or if this is something you might consider useful, but it could allow such syntax:

```ts
import { createAdapter, joinAdapters } from '@state-adapt/core';
import { createCollectionAdapter} from '@state-adapt/core/adapters';

export interface TodoItem {
  id: number;
  title: string;
  completed: boolean;
}

export interface TodoItemsState {
  pagination: Pagination;
  todoItems: TodoItem[];  // 👈 Native types!
}

export const todoItemsStateAdapter = joinAdapters<TodoItemsState>()({
  pagination: paginationAdapter,
  todoItems: createCollectionAdapter<TodoItem>(),
})();

const initialState: Readonly<TodoItemsState> = {
  pagination: { /*...*/ },
  todoItems: [],  // 👈 Native collections!
};
```

> Note: this won't rely on a previous `todoItemAdapter` that might have been defined, nor does is take another parameter to add behaviour, that could be useful however

Would it be ok if I'm drafting a PR about this?

**Comment by mfp22** [comment]
I've thought a lot about this. 

The entity adapter is annoying. For NgRx and everything. I would often love to just use an array. And sometimes I do just so I don't have to jump through the entity hoops. 

The reason for the entity adapter (in NgRx too) is the shared utilities, but also the efficiency of how the state is structured. If you are getting an entity by id... entities[id] is fast... whereas items.find(item => ...) is... actually probably fast enough for the vast majority of cases. How often are people going to be storing 1000+ entities? Probably not often.

Alright, I'm convinced. The entity adapter is _fine_ for simple cases, but not ideal. It's good for performance benchmarks. 

For the implementation, I think 1 of 2 approaches would be best: Extremely similar to native `Array` methods, or to the entity adapter state change functions. The reason is because `Array` methods are familiar to everyone, and StateAdapt's philosophy is incremental adaptation to complexity, so if performance becomes a big concern, someone should be able to switch to the entity adapter without rewriting everything. 

Okay, so here's what I think. We should make it similar to `Array` methods. Most people are never going to use the entity adapter if this exists. It really is less friction to use an array adapter like you mentioned. It will be easier to remember. I can't even remember how to use the entity adapter. 

I would just call it `createArrayAdapter`. I prefer shorter names, and `collection` might sound to some people like it has extra structure, like the entity adapter.

But I still want the composability of the entity adapter. So you could do `createArrayAdapter(counterAdapter)` and get `incrementAll`. And support for `filters: []` and `sorters: []`.

The difference will be that the `id` selector should probably use the index in the array. There might be some other subtle differences.

I think it would make sense to implement an alpha version and warn people before using it. When moving away from something as well-designed as the entity adapter (RTK even copied it from NgRx), I want to be careful that I'm not ignoring something important.

**Reply by pBouillon** [reply]
I agree with `array` instead of `collection`, you're right and it might confuse people regarding the data structure it represents

However, things are not perfectly clear on my side regarding the API of this function.

Let's consider this setup:

```ts
interface Item { name: string; price: number; isSoldOut: boolean }

const items = [
  { name: 'Milk', price: 4.99, isSoldOut: false },
  { name: 'Croissant', price: 3.50, isSoldOut: true },
];
```

Does something like this matches what you have in mind?

```ts
// Adapter for the element
const itemAdapter = createAdapter<Item>()({ /*...*/});

// Adapter for the elements based on the previous adapter
const itemsAdapter = createArrayAdapter<Item>(itemAdapter, {
  //  👇 Additional behavior based on the keys of the `TItem` of `createArrayAdapter`
  filter: ['isSoldOut'],
  sorter: ['price'] ],
});

const store = adapt(items, {  // 👈 Built directly with the "native" value
  adapter: itemsAdapter,
  sources: /*...*/,
});
```

Personally, I think that having the option to use `createArrayAdapter` as a simple adapter might encourage its adoption for use cases when having to much feature is not an advantage.

Consider this:

```ts
const featuredItemsStore = adapt(new Array<Item>(), { 
  adapter: createArrayAdapter<Item>(),
  sources: /*...*/,
});
```

In this use case, or in a similar one, I might only want to manage the array itself: get all elements, clear all of them, get the total count, etc.

While having the opportunity to enhance the behavior of the adapter by composing it with an adapter of the array items, making it mandatory might introduce friction and/or confusion in its usage.

As for the resulting adapter itself, do you have an API in mind for its selectors, reaction, etc.?
I'm not sure I am familiar enough with TS type system to submit a PR with this function but I can sure try to get started and iterate with you on that if you want some help!

**Reply by mfp22** [reply]
Yeah the inner adapter can be optional. I thought I made it optional for the entity adapter too. Didn't I? I don't remember.

I think the implementation itself can be pretty similar to the entity adapter.

**Reply by mfp22** [reply]
Looks like I didn't make it optional. It easily could be though. I think you can probably just pass `{}` into it at least.
